hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/AbstractRolloutExecutor.java

```diff
@@ -0,0 +1,236 @@
+                        retries, rollout.getId(), rollout.getName(), e);
+                handleErrorStarting(jpaRollout);
+            });
+            break;
+        case RUNNING:
+            handleRunningRollout(jpaRollout);
+            break;
+        case STOPPING:
+            handleStopRollout(jpaRollout);
+            break;
+        default:
+            LOGGER.error("Rollout in status {} not supposed to be handled!", rollout.getStatus());
+            break;
+        }
+    }
+
+    protected abstract void handleCreateRollout(final JpaRollout rollout);
+
+    protected abstract void handleDeleteRollout(final JpaRollout rollout);
+
+    protected abstract void handleReadyRollout(final JpaRollout rollout);
+
+    protected abstract void handleStartingRollout(final JpaRollout rollout);
+
+    protected abstract void handleRunningRollout(final JpaRollout rollout);
+
+    protected abstract void handleStopRollout(final JpaRollout rollout);
+
+    protected void handleErrorCreating(final JpaRollout rollout) {
+        setRolloutErrorState(rollout);
+    }
+
+    protected void handleErrorStarting(final JpaRollout rollout) {
+        setRolloutErrorState(rollout);
+    }
+
+    protected void handleErrorDeleting(final JpaRollout rollout) {
+        setRolloutErrorState(rollout);
+    }
+
+    private void setRolloutErrorState(final JpaRollout rollout) {
+        if (!cleanupScheduledActions(rollout)) {
+            // could not clean up scheduled actions
+            LOGGER.warn("Could not clean up scheduled actions for rollout with id {}", rollout.getId());
+        }
+        setErrorStateForRolloutGroups(rollout);
+        rollout.setStatus(Rollout.RolloutStatus.ERROR_STARTING);
+        rolloutRepository.save(rollout);
+    }
+
+    /**
+     * @param rollout
+     * @return if all scheduled actions could be stopped
+     */
+    protected boolean cleanupScheduledActions(final JpaRollout rollout) {
+        // clean up all scheduled actions
+        deleteScheduledActions(rollout);
+
+        // avoid another scheduler round and re-check if all scheduled actions
+        // has been cleaned up. we flush first to ensure that the we include the
+        // deletion above
+        entityManager.flush();
+        return actionRepository.countByRolloutIdAndStatus(rollout.getId(), Action.Status.SCHEDULED) == 0;
+    }
+
+    protected void deleteScheduledActions(final JpaRollout rollout) {
+        final Slice<JpaAction> scheduledActions = findScheduledActionsByRollout(rollout);
+        final boolean hasScheduledActions = scheduledActions.getNumberOfElements() > 0;
+
+        if (hasScheduledActions) {
+            try {
+                final Iterable<JpaAction> iterable = scheduledActions::iterator;
+                final List<Long> actionIds = StreamSupport.stream(iterable.spliterator(), false).map(Action::getId)
+                        .collect(Collectors.toList());
+                actionRepository.deleteByIdIn(actionIds);
```

Comment:
please consider splitting action ids in the batches of 999(max entries in IN statement), because the TRANSACTION_ACTIONS is 5000